Scripting API documentation (#14725)
mdn/content: files/en-us/mozilla/add-ons/webextensions/api/scripting/executescript/index.md

---
title: scripting.executeScript()
slug: Mozilla/Add-ons/WebExtensions/API/scripting/executeScript
tags:
  - API
  - Add-ons
  - Extensions
  - Method
  - Reference
  - WebExtensions
  - executeScript
  - scripting
browser-compat: webextensions.api.scripting.executeScript
---
{{AddonSidebar()}}

Injects a script into a target context. The script is run at `document_idle` by default.

> **Note:** This method is available in Manifest V3 or higher.

> The scripts you inject are called [content scripts](/en-US/docs/Mozilla/Add-ons/WebExtensions/Content_scripts).

Reviewer: ```suggestion
 To use this API you must have the `"scripting"` [permission](/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/permissions) and permission for the page's URL, either explicitly as a [host permission](/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/permissions#host_permissions) or using the [activeTab permission](/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/permissions#activetab_permission).
 
The scripts you inject are called [content scripts](/en-US/docs/Mozilla/Add-ons/WebExtensions/Content_scripts).
```

---
title: scripting.executeScript()
slug: Mozilla/Add-ons/WebExtensions/API/scripting/executeScript
tags:
  - API
  - Add-ons
  - Extensions
  - Method
  - Reference
  - WebExtensions
  - executeScript
  - scripting
browser-compat: webextensions.api.scripting.executeScript
---
{{AddonSidebar()}}

Injects a script into a target context. The script is run at `document_idle` by default.

> > **Note:** This method is available in Manifest V3 or higher.

Author: ```suggestion
> **Note:** This method is available in Manifest V3 or higher. It's also available in Manifest V2 from Firefox version 102.
```

## Syntax

```js
let results = await browser.scripting.executeScript(
  details             // object
)
```

### Parameters

- `details`

  - : An object describing the script to inject.  It contains the following properties:

    - `args`
      - : `any`. An array of arguments to carry into the function. This is only valid if the `func` parameter is specified. The arguments must be JSON-serializable.
    - `files`
      - : `string`. An array of path of the JS files to inject, relative to the extension's root directory. Exactly one of `files` and `func` must be specified.
    - `func`
      - : `function`. A JavaScript function to inject. This function is serialized and then deserialized for injection. This means that any bound parameters and execution context ase lost. Exactly one of `files` and `func` must be specified.
    - `injectImmediately`{{optional_inline}}
      - : `boolean`. Whether the injection into the target is triggered as soon as possible, but not necessarily prior to page load.
    - `target`
      - : {{WebExtAPIRef("scripting.InjectionTarget")}}. Details specifying the target to inject the script into.   

### Return value

> A [`Promise`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise) that fulfills with an array of {{WebExtAPIRef("scripting.InjectionResult")}}. The array's values represent the result of the script in every injected frame. If any error occurs, the promise is rejected.

Reviewer: ```suggestion
A [`Promise`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise) that fulfills with an array of `InjectionResult` objects, which represent the result of the injected script in every injected frame. If any error occurs, the promise is rejected.

Each `InjectionResult` object has the following properties:

- `frameId`
  - : `number`. The frame ID associated with the injection.
- `result`{{optional_inline}}
  - : `any`. The result of the script execution.
- `error`{{optional_inline}}
  - : `object`. When the injection fails, details of the failure errors.
    - `message`
      - : `string`. A message explaining why the injection failed.
```

---
title: scripting.executeScript()
slug: Mozilla/Add-ons/WebExtensions/API/scripting/executeScript
tags:
  - API
  - Add-ons
  - Extensions
  - Method
  - Reference
  - WebExtensions
  - executeScript
  - scripting
browser-compat: webextensions.api.scripting.executeScript
---
{{AddonSidebar()}}

Injects a script into a target context. The script is run at `document_idle` by default.

> > **Note:** This method is available in Manifest V3 or higher. It's also available in Manifest V2 from Firefox version 102.

Reviewer: About this method being supported in MV2 from Firefox version 102. This is true for everything that we documented here, the whole `scripting` namespace is available in FF 102+.

Author: I've updated all the notes,

Author: I went for the note update, checked the BCD and eliminated the Safari reference where appropriate.

---
title: scripting.executeScript()
slug: Mozilla/Add-ons/WebExtensions/API/scripting/executeScript
tags:
  - API
  - Add-ons
  - Extensions
  - Method
  - Non-standard
  - Reference
  - WebExtensions
  - executeScript
  - scripting
browser-compat: webextensions.api.scripting.executeScript
---
{{AddonSidebar()}}

Injects a script into a target context. The script is run at `document_idle`.

> **Note:** This method is available in Manifest V3 or higher.

The scripts you inject are called [content scripts](/en-US/docs/Mozilla/Add-ons/WebExtensions/Content_scripts).

This is an asynchronous function that returns a [`Promise`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise).

## Syntax

```js
let executing = browser.scripting.executeScript(
  injection,             // object
  callback               // function
)
```

### Parameters

- `injection`
  - : {{WebExtAPIRef("scripting.ScriptInjection")}}. Details of a script to inject.
- `callback`{{optional_inline}} 
  - : `function`. Invoked upon completion of the request.

### Return value

> A [`Promise`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise) that is fulfilled with {{WebExtAPIRef("scripting.InjectionResult")}}. The array's values represent the result of the script in every injected frame.

Author: @Rob--W I don't follow, so this doesn't return InjectionResult? Also, I'm struggling to find an example where content returned from a promise is in line. Are you aware of any examples?

> The result of the script is the last evaluated statement, which is similar the results that would be seen if you executed the script in the [Web Console](/en-US/docs/Tools/Web_Console) (not any `console.log()` output). For example, consider a script like this:

Reviewer: ```suggestion
The result of the script is the last evaluated statement, which is similar to the results that would be seen if you executed the script in the [Web Console](/en-US/docs/Tools/Web_Console) (not any `console.log()` output). For example, consider a script like this:
```

Reviewer: > (not any `console.log()` output)

What does that mean?

```js
let foo='my result'; foo;
```

Here the results array contains the string "`my result`" as an element.

The result values must be [structured clonable](/en-US/docs/Web/API/Web_Workers_API/Structured_clone_algorithm) (see [Data cloning algorithm](/en-US/docs/Mozilla/Add-ons/WebExtensions/Chrome_incompatibilities#data_cloning_algorithm)).

> If any error occurs, the promise is rejected with an error message.

Author: @Rob--W could you be more explicit about what you think should be said about the era returned?

## Examples

This example executes a one-line code snippet in the active tab:

```js
function onExecuted(result) {
  console.log(`We made it green`);
}

function onError(error) {
  console.log(`Error: ${error}`);
}

const makeItGreen = 'document.body.style.border = "5px solid green"';

> const executing = browser.tabs.executeScript({

Reviewer: Shouldn't we document the `scripting` namespace?

  code: makeItGreen
});

> executing.then(onExecuted, onError);

Reviewer: async/await would make the code more readable IMO but I don't know what the guidelines are.

```

This example executes a script from a file (packaged with the extension) called `"content-script.js"`. The script is executed in the active tab. The script is executed in subframes and the main document:

```js
function onExecuted(result) {
  console.log(`We executed in all subframes`);
}

function onError(error) {
  console.log(`Error: ${error}`);
}

const executing = browser.tabs.executeScript({

>   file: "/content-script.js",

Reviewer: I don't think we support `file`, instead it should be:

```suggestion
  files: ["content-script.js"],
```

Reviewer: Also I think the `target` is mandatory.
